helix-core/src/main/java/org/apache/helix/manager/zk/CallbackEventTPFactory.java

```diff
@@ -0,0 +1,76 @@
+package org.apache.helix.manager.zk;
+
+
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+// returns a thread pool object given a HelixManager identifier
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+
+
+public class CallbackEventTPFactory {
+  static private final ReadWriteLock _lock = new ReentrantReadWriteLock();
+  static private Map<Integer, ThreadPoolExecutor> _managerToCallBackThreadPoolMap = new HashMap();
+
+  public static ThreadPoolExecutor getThreadPool(int mapHash) {
+    // should not use general lock for read
+    _lock.readLock().lock();
+    ThreadPoolExecutor result;
+    if (_managerToCallBackThreadPoolMap.containsKey(mapHash)) {
+      result = _managerToCallBackThreadPoolMap.get(mapHash);
+      _lock.readLock().unlock();
+    } else {
+      _lock.readLock().unlock();
+      result = getAndCreateThreadPool(mapHash);
+    }
+    return result;
+  }
+
+  public static ThreadPoolExecutor getAndCreateThreadPool(int mapHash) {
```

Comment:
Why we have 2 public methods that are doing. almost the same thing?
I would suggest having one for getOrCreate(), another one for create() with writelock.